src/data_generators/data_set.py

```diff
@@ -0,0 +1,192 @@
+        if len(impressions) != self.publisher_count:
+            raise ValueError(
+                "Invalid impression vector length.  Got {}, expected {}".format(
+                    len(impressions), self.publisher_count
+                )
+            )
+        counts = defaultdict(int)
+        spends = []
+        for i, imp in enumerate(impressions):
+            spends.append(self._data_files[i].spend_by_impressions(imp))
+            for id, freq in self._data_files[i].user_counts_by_impressions(imp).items():
+                counts[id] += freq
+        kplus_reaches = self._counts_to_histogram(counts, max_frequency)
+        return ReachPoint(impressions, kplus_reaches, spends)
+
+    def _counts_to_histogram(
+        self, counts: Dict[int, int], max_frequency: int
+    ) -> List[int]:
+        """Constructs k+ reach list from a dictionary of per-id reach counts."""
+        frequency_counts = [0] * max_frequency
+        for c in counts.values():
+            frequency_counts[min(c, max_frequency) - 1] += 1
+        # At this point, frequency_counts[k] represents the number of people who are
+        # reach exactly k+1 times, except that frequency_counts[max_frequency-1] contains
+        # the number of people reached at least max_frequency times.  Now, we convert this
+        # to a list of k+ reach values.
+        for i in range(max_frequency - 2, -1, -1):
+            frequency_counts[i] += frequency_counts[i + 1]
+        return frequency_counts
+
+    def reach_by_spend(
+        self, spends: Iterable[float], max_frequency: int = 10
+    ) -> ReachPoint:
+        """Number of people reached for a given spend.
+
+        Args:
+          spends:  A list of spend amounts.  The length of the list must
+            equal the value of publisher_count.  Specifies the amount spent with
+            each publisher.
+          max_frequency: int, The maximum frequency that should be counted.  All
+            frequencies about this amount will be grouped into a single bucket.
+        Returns:
+          A ReachPoint object representing the k+ reach for each frequency
+          in the range 1..max_frequency.
+        """
+        if len(spends) != self.publisher_count:
+            raise ValueError(
+                "Invalid spends vector length.  Got {}, expected {}".format(
+                    len(spends), self.publisher_count
+                )
+            )
+        counts = defaultdict(int)
+        impressions = []
+        for i, publisher_spend in enumerate(spends):
+            user_counts = self._data_files[i].user_counts_by_spend(publisher_spend)
+            impressions.append(sum(user_counts.values()))
+            for id, freq in user_counts.items():
+                counts[id] += freq
+        kplus_reaches = self._counts_to_histogram(counts, max_frequency)
+        return ReachPoint(impressions, kplus_reaches, spends)
+
+    def write_data_set(self, parent_dir: str, dataset_dir: str = None) -> None:
```

Comment:
Consider passing file object (as below).

Reply:
A file object won't work in this case, because what is being passed is a directory name, in which a number of files are being created.  However, I modified the PublisherData class so that it uses file objects.